src/azure-cli/azure/cli/command_modules/util/custom.py

```diff
@@ -188,28 +188,33 @@ def _upgrade_on_windows():
     This also gives the user a chance to manually install the MSI in case of msiexec.exe failure.
     """
     import platform
+    import subprocess
+    import sys
+    import tempfile
+
+    from pathlib import Path
+    from azure.cli.core.util import rmtree_with_retry
 
     if platform.architecture()[0] == '32bit':
         msi_url = 'https://aka.ms/installazurecliwindows'
     else:
         msi_url = 'https://aka.ms/installazurecliwindowsx64'
     logger.warning("Updating Azure CLI with MSI from %s", msi_url)
-    tmp_dir, msi_path = _download_from_url(msi_url)
 
-    logger.warning("Installing MSI")
-    import subprocess
-    exit_code = subprocess.call(['msiexec.exe', '/i', msi_path])
+    # Save MSI to ~\AppData\Local\Temp\azure-cli-msi, clean up the folder first
+    msi_dir = Path(tempfile.gettempdir()) / 'azure-cli-msi'
+    rmtree_with_retry(msi_dir)
+    msi_dir.mkdir()
```

Comment:
Will this line raise exception if `rmtree_with_retry` can't delete the directory successfully?

Reply:
`rmtree_with_retry` ignores the `FileNotFoundError` by default, which is my desired behavior.

Reply:
`rmtree_with_retry` also ingores cannot-delete-folder error, and that's why the directory can still exist and cause error to `mkdir`.